CA02: This is a eMail Spam Classifers that uses Naive Bayes supervised machine learning algorithm. 

In this assignment you will ...
1. Complete the code such a way that it works correctly with this given parts of the program.
2. Explain as clearly as possible what each part of the code is doing. Use "Markdown" texts and code commenting to explain the code

IMPORTANT NOTE:

The path of your data folders 'train-mails' and 'test-mails' must be './train-mails' and './test-mails'. This means you must have your .ipynb file and these folders in the SAME FOLDER in your laptop or Google Drive. The reason for doing this is, this way the peer reviewes and I would be able to run your code from our computers using this exact same relative path, irrespective of our folder hierarchy.

In [1]:
import os
import numpy as np
from collections import Counter

# Import all other necessary libraries. Your code below ...
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
import pandas as pd


Make Dictionary function to give the most common 3000 words

In [2]:
'''This function builds a Dictionary of most common 3000 words from all the email content. 
First it adds all words and symbols in the dictionary. 
Then it removes all non-alpha-numeric characters and any single character alpha-numeric characters. 
After this is complete it shrinks the Dictionary by keeping only most common 3000 words in the dictionary. '''

def make_Dictionary(root_dir):
  all_words = []
  emails = [os.path.join(root_dir,f) for f in os.listdir(root_dir)]
  for mail in emails:
    with open(mail) as m:
      for line in m:
        words = line.split()
        all_words += words
  dictionary = Counter(all_words)
  list_to_remove = list(dictionary)

  for item in list_to_remove:
    if item.isalpha() == False:
      del dictionary[item]
    elif len(item) == 1:
      del dictionary[item]
  dictionary = dictionary.most_common(3000)
  return dictionary
            

Creating feature columns and labels

In [3]:
"""This function extracts feature columns and populates their values 
(Feature Matrix of 3000 comumns and rows equal to the number of email files). 
The function also analyzes the File Names of each email file and decides if it's 
a Spam or not based on the naming convention. The function also 
creates the Labelled Data Column. This function is used to extract the training dataset 
as well as the testing dataset and returns the Feature Dataset and the Label column."""

def extract_features(mail_dir):
  files = [os.path.join(mail_dir,fi) for fi in os.listdir(mail_dir)]
  features_matrix = np.zeros((len(files),3000))
  train_labels = np.zeros(len(files))
  count = 1;
  docID = 0;
  for fil in files:
    with open(fil) as fi:
      for i, line in enumerate(fi):
        if i ==2:
          words = line.split()
          for word in words:
            wordID = 0
            for i, d in enumerate(dictionary):
              if d[0] == word:
                wordID = i
                features_matrix[docID,wordID] = words.count(word)
      train_labels[docID] = 0;
      filepathTokens = fil.split('/')
      lastToken = filepathTokens[len(filepathTokens)-1]
      if lastToken.startswith("spmsg"):
        train_labels[docID] = 1;
        count = count + 1
      docID = docID + 1
  return features_matrix, train_labels                

In [4]:
# Enter the "path" of your "train_mails" and "test-mails" FOLDERS in this cell ...
TRAIN_DIR = '/content/drive/My Drive/Data/train-mails'
TEST_DIR = '/content/drive/My Drive/Data/test-mails'

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


This step is creating a training dictionary

In [6]:
dictionary = make_Dictionary(TRAIN_DIR)

print ("reading and processing emails from TRAIN and TEST folders")
features_matrix, labels = extract_features(TRAIN_DIR)
test_features_matrix, test_labels = extract_features(TEST_DIR)

reading and processing emails from TRAIN and TEST folders


In [7]:
"""The section is the main Program that calls the two functions above. 
First it "trains" the model using model.fit function and Training Dataset. 
After that it scores the Test Data set by running the Trained Model with the Test Data set. 
At the end it prints the model performance in terms of accuracy score."""

model = GaussianNB()

print ("Training Model using Gaussian Naibe Bayes algorithm .....")
model.fit(features_matrix, labels)
print ("Training completed")
print ("testing trained model to predict Test Data labels")
predicted_labels = model.predict(test_features_matrix)
print ("Completed classification of the Test Data .... now printing Accuracy Score by comparing the Predicted Labels with the Test Labels:")
print (accuracy_score(test_labels, predicted_labels))

Training Model using Gaussian Naibe Bayes algorithm .....
Training completed
testing trained model to predict Test Data labels
Completed classification of the Test Data .... now printing Accuracy Score by comparing the Predicted Labels with the Test Labels:
0.9653846153846154


The results of the code show the features being extracted using the model, the model being trained by the functions, the classification of the data, and finally the accuracy, which is extremely high, at 96.5%

======================= END OF PROGRAM =========================